spring-xd-dirt/src/main/java/org/springframework/xd/dirt/batch/tasklet/JobLaunchingTasklet.java

```diff
@@ -0,0 +1,273 @@
+import org.springframework.util.Assert;
+import org.springframework.util.StringUtils;
+import org.springframework.xd.dirt.integration.bus.MessageBus;
+import org.springframework.xd.dirt.stream.JobDefinition;
+import org.springframework.xd.dirt.stream.JobDefinitionRepository;
+import org.springframework.xd.dirt.stream.NoSuchDefinitionException;
+import org.springframework.xd.dirt.stream.NotDeployedException;
+import org.springframework.xd.store.DomainRepository;
+
+/**
+ * A {@link Tasklet} implementation that uses the Spring XD {@link MessageBus} to launch
+ * jobs deployed within the same Spring XD cluster.  This tasklet also receives the
+ * results of the job from the same message bus.
+ *
+ * The result of the step executing this tasklet should match that of the job that was
+ * executed by this job.  So if the tasklet executes the job foo and foo returns an
+ * {@link org.springframework.batch.core.ExitStatus} of "BAR", the ExitStatus of this step
+ * will also be "BAR".
+ *
+ * @author Michael Minella
+ * @since 1.3.0
+ */
+public class JobLaunchingTasklet implements Tasklet, MessageHandler {
+
+	public static final String XD_ORCHESTRATION_ID = "xd_orchestration_id";
+
+	//TODO: Make this configurable
+	private long timeout = -1;
+
+	//TODO: Make this configurable
+	private long pollInterval = 1000;
+
+	private String jobName;
+
+	private MessageBus messageBus;
+
+	private volatile JobExecution results = null;
+
+	private JobDefinitionRepository definitionRepository;
+
+	private DomainRepository<JobDefinition, String> instanceRepository;
+
+	private String orchestrationId;
+
+	private JobParametersExtractor extractor = new JobParametersExtractor();
+
+	private MessageChannel launchingChannel;
+
+	private PublishSubscribeChannel listeningChannel;
+
+	public JobLaunchingTasklet(MessageBus messageBus,
+			JobDefinitionRepository jobDefinitionRepository,
+			DomainRepository instanceRepository, String jobName) {
+		this(messageBus, jobDefinitionRepository, instanceRepository, jobName,
+				new DirectChannel(),
+				new PublishSubscribeChannel((new SimpleAsyncTaskExecutor())));
+	}
+
+	/**
+	 * Provided for testing to be able to inject the channels for mocking.
+	 *
+	 * @param messageBus Message bus reference to launch a job and receive it's results
+	 * @param jobDefinitionRepository Repository used to look up the child job definition
+	 * @param instanceRepository Repository used to look up that the child job is deployed
+	 * @param jobName The name of the child job definition
+	 * @param launchingChannel The channel used to send the launch request
+	 * @param listeningChannel The channel used to listen for the job results
+	 */
+	protected JobLaunchingTasklet(MessageBus messageBus,
+			JobDefinitionRepository jobDefinitionRepository,
+			DomainRepository instanceRepository, String jobName,
+			MessageChannel launchingChannel, PublishSubscribeChannel listeningChannel) {
+		Assert.notNull(messageBus, "A message bus is required.");
```

Comment:
Assertion message consistency - Personally I favor "full stops" but we should not mix.